src/datanode/src/region_server.rs

```diff
@@ -450,6 +465,33 @@ struct RegionServerInner {
     runtime: Runtime,
     event_listener: RegionServerEventListenerRef,
     table_provider_factory: TableProviderFactoryRef,
+    // The number of queries allowed to be executed at the same time.
+    // Act as last line of defense on datanode to prevent query overloading.
+    parallelism: Option<RegionServerParallelism>,
+}
+
+struct RegionServerParallelism {
+    semaphore: Semaphore,
+    timeout: Duration,
+}
+
+impl<'a> RegionServerParallelism {
+    pub fn from_opts(max_concurrent_queries: usize, timeout: Duration) -> Option<Self> {
+        if max_concurrent_queries == 0 {
+            return None;
+        }
+        Some(RegionServerParallelism {
+            semaphore: Semaphore::new(max_concurrent_queries),
+            timeout,
+        })
+    }
+
+    pub async fn acquire(&'a self) -> Result<SemaphorePermit<'a>> {
```

Comment:
Looks like we can remove the lifetime parameter
```suggestion
impl RegionServerParallelism {
    pub fn from_opts(max_concurrent_queries: usize, timeout: Duration) -> Option<Self> {
        if max_concurrent_queries == 0 {
            return None;
        }
        Some(RegionServerParallelism {
            semaphore: Semaphore::new(max_concurrent_queries),
            timeout,
        })
    }

    pub async fn acquire(&self) -> Result<SemaphorePermit> {
```